cleverhans/picklable_model.py

```diff
@@ -0,0 +1,370 @@
+            for i, layer in enumerate(self.layers[:-1]):
+                if layer.name == name:
+                    layers = self.layers[i+1:]
+                    break
+
+        for layer in layers:
+            x = out[layer.parent]
+            if set_ref:
+                layer.ref = x
+            x = layer.fprop(x)
+            assert x is not None
+            out[layer.name] = x
+        return out
+
+
+class Layer(PicklableModel):
+    def __init__(self, name=None, parent=None):
+        if name is not None:
+            self.name = name
+        self.parent = parent
+
+    def get_output_shape(self):
+        return self.output_shape
+
+
+class Linear(Layer):
+
+    def __init__(self, num_hid, init_scale=1., init_b=0., **kwargs):
+        super(Linear, self).__init__(**kwargs)
+        self.num_hid = num_hid
+        self.init_scale = init_scale
+        self.init_b = init_b
+
+    def set_input_shape(self, input_shape):
+        batch_size, dim = input_shape
+        self.input_shape = [batch_size, dim]
+        self.output_shape = [batch_size, self.num_hid]
+        init = tf.random_normal([dim, self.num_hid], dtype=tf.float32)
+        init = init / tf.sqrt(1e-7 + tf.reduce_sum(tf.square(init), axis=0,
+                                                   keep_dims=True))
+        init = init * self.init_scale
+        self.W = PV(init)
+        self.b = PV((np.zeros((self.num_hid,))
+                     + self.init_b).astype('float32'))
+
+    def fprop(self, x):
+        return tf.matmul(x, self.W.var) + self.b.var
+
+    def get_params(self):
+        return [self.W.var, self.b.var]
+
+
+class Conv2D(Layer):
+
+    def __init__(self, output_channels, kernel_shape, strides, padding,
+                 use_bias=False, init_scale=1., **kwargs):
+        self.__dict__.update(locals())
+        del self.self
+        super(Conv2D, self).__init__(**kwargs)
+
+    def set_input_shape(self, input_shape):
+        batch_size, rows, cols, input_channels = input_shape
+        kernel_shape = tuple(self.kernel_shape) + (input_channels,
+                                                   self.output_channels)
+        assert len(kernel_shape) == 4
+        assert all(isinstance(e, int) for e in kernel_shape), kernel_shape
+        init = tf.random_normal(kernel_shape, dtype=tf.float32)
+        init = self.init_scale * init / tf.sqrt(1e-7 +
+                                                tf.reduce_sum(tf.square(init),
+                                                              axis=(0, 1, 2)))
+        self.kernels = PV(init)
+        if self.use_bias:
+            self.b = PV(np.zeros((self.output_channels,)).astype('float32'))
+        orig_input_batch_size = input_shape[0]
+        input_shape = list(input_shape)
+        input_shape[0] = 1
+        dummy_batch = tf.zeros(input_shape)
+        dummy_output = self.fprop(dummy_batch)
+        output_shape = [int(e) for e in dummy_output.get_shape()]
+        output_shape[0] = 1
```

Comment:
Was this meant to be: `output_shape[0] = orig_input_batch_size`?

Reply:
Yes, thanks. Fixed.